# Liabrary

In [27]:
import pandas as pd
import numpy as np

#Statistic
from scipy import stats

# Visualization
import plotly.express as px
import matplotlib.pyplot as plt

# machine learning
import statsmodels.api as sm

# Dataframe

In [3]:
df=pd.read_csv("C:\\Users\\Saravanan\\OneDrive\\Desktop\\Datasets_data\\Cars - Purchase Decision.csv")

In [4]:
df

User ID  Gender  Age  AnnualSalary  Purchased
0        385    Male   35         20000          0
1        681    Male   40         43500          0
2        353    Male   49         74000          0
3        895    Male   40        107500          1
4        661    Male   25         79000          0
..       ...     ...  ...           ...        ...
995      863    Male   38         59000          0
996      800  Female   47         23500          0
997      407  Female   28        138500          1
998      299  Female   48        134000          1
999      687  Female   44         73500          0

[1000 rows x 5 columns]

# Observation

Depedent     -  purchased
Indepent     -  AnnualSalary, Gender,	Age

catagory     - Gender,Purchased
continous    - age,annual salary

# supervised classification

# Data cleaning

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 5 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   User ID       1000 non-null   int64 
 1   Gender        1000 non-null   object
 2   Age           1000 non-null   int64 
 3   AnnualSalary  1000 non-null   int64 
 4   Purchased     1000 non-null   int64 
dtypes: int64(4), object(1)
memory usage: 39.2+ KB


In [6]:
df.isnull().sum()

User ID         0
Gender          0
Age             0
AnnualSalary    0
Purchased       0
dtype: int64

In [7]:
df.set_index("User ID",inplace=True)

In [8]:
df.head()

Gender  Age  AnnualSalary  Purchased
User ID                                     
385       Male   35         20000          0
681       Male   40         43500          0
353       Male   49         74000          0
895       Male   40        107500          1
661       Male   25         79000          0

# Statistical analysis

Continuous - Central Limit Theorem, 1-Tailed Test (one-sample t-test)

Continuous vs. Continuous - Correlation, 2-Tailed Test (two-sample t-test)

Continuous vs. Categorical - ANOVA (Analysis of Variance)

Categorical vs. Categorical - Chi-Square Test

# Hypothesis Testing

Continuous - Age, AnnualSalary
Category - Gender, Purchased

# Central limit theoram

Age - population_mean : 40.106 sample_mean : 40.1015

Annual - population_mean : 72689.0 sample_mean : 72425.16666666667

# 1 tailed test
age- H0 is rejected, Ha is accepted, There is a significant effect

Annual - H0 is rejected, Ha is accepted, There is a significant effect

# 2 tailed test 
H0 is rejected, There is no relationship between two columns we're comparing.

# Chi sqi test
H0 is accepted, There is no relationship between two columns we're comparing

# Anova
Gender,age - H0 is rejected , There is no relationship between two columns we're comparing

Gender,annual- H0 is rejected , There is no relationship between two columns we're comparing

purchased,age - H0 is rejected , There is no relationship between two columns we're comparing

purchased,annual - H0 is rejected , There is no relationship between two columns we're comparing

# Correlation

In [9]:
df.corr(numeric_only=True)


Age  AnnualSalary  Purchased
Age           1.000000      0.166042   0.616036
AnnualSalary  0.166042      1.000000   0.364974
Purchased     0.616036      0.364974   1.000000

# Cental Limit Theoram

In [10]:
population=df.Age.values
population_mean=population.mean()
samplesize=150
sample_mean=[]
for i in range(40):
    sample=np.random.choice(population,samplesize)
    sample_mean.append(sample.mean())
np.mean(sample_mean)
print("population_mean :",population_mean,"sample_mean :",np.mean(sample_mean))

population_mean : 40.106 sample_mean : 40.19716666666667


In [11]:
population=df.AnnualSalary.values
population_mean=population.mean()
samplesize=150
sample_mean=[]
for i in range(40):
    sample=np.random.choice(population,samplesize)
    sample_mean.append(sample.mean())
np.mean(sample_mean)
print("population_mean :",population_mean,"sample_mean :",np.mean(sample_mean))

population_mean : 72689.0 sample_mean : 72912.25


# 1 Tailed Test

In [12]:
H0_accepted = 0
H0_rejected = 0
for i in range(30):
  sample = df.Age.sample(frac=0.05)
  t_test, p_value = stats.ttest_1samp(sample,df.Age.mean())
  if p_value > 0.5:
    H0_accepted += 1
  else:
    H0_rejected += 1

print(f"H0 accepted {H0_accepted} times , H0 rejected {H0_rejected} times")
if H0_accepted > H0_rejected:
  print("H0 is accepted, Ha is rejected, There is no significant effect")
else:
  print("H0 is rejected, Ha is accepted, There is a significant effect")

H0 accepted 15 times , H0 rejected 15 times
H0 is rejected, Ha is accepted, There is a significant effect


In [13]:

H0_accepted = 0
H0_rejected = 0
for i in range(30):
  sample = df.AnnualSalary.sample(frac=0.05)
  t_test, p_value = stats.ttest_1samp(sample,df.AnnualSalary.mean())
  if p_value > 0.5:
    H0_accepted += 1
  else:
    H0_rejected += 1

print(f"H0 accepted {H0_accepted} times , H0 rejected {H0_rejected} times")
if H0_accepted > H0_rejected:
  print("H0 is accepted, Ha is rejected, There is no significant effect")
else:
  print("H0 is rejected, Ha is accepted, There is a significant effect")

H0 accepted 13 times , H0 rejected 17 times
H0 is rejected, Ha is accepted, There is a significant effect


# 2 Tailed Test

In [14]:
H0_accepted = 0
H0_rejected = 0
for i in range(20):
  sample1 = df.Age.sample(frac=0.04)
  sample2 = df.AnnualSalary.sample(frac=0.04)
  t_test, p_value = stats.ttest_ind(sample1, sample2)
  if p_value > 0.5:
    H0_accepted += 1
  else:
    H0_rejected += 1

print(f"H0 accepted {H0_accepted} times , H0 rejected {H0_rejected} times")
if H0_accepted > H0_rejected:
  print("H0 is accepted, There is a relationship between two columns we're comparing")
else:
  print("H0 is rejected, There is no relationship between two columns we're comparing.")


H0 accepted 0 times , H0 rejected 20 times
H0 is rejected, There is no relationship between two columns we're comparing.


# Chi Square Test

In [15]:

data = pd.crosstab(df.Gender, df.Purchased)
observed_values = data.values
chi2_stat, p_value, _, _ = stats.chi2_contingency(observed_values)
if p_value > 0.05:
  print("H0 is accepted, There is no relationship between two columns we're comparing")
else:
  print("H0 is rejected, There is a relationship between two columns we're comparing")

H0 is accepted, There is no relationship between two columns we're comparing


# Anova

In [16]:
group=df.Gender.unique()
data={}
for i in group:
    data[i]=df.Age[df.Gender==i]
f_value,p_value=stats.f_oneway(*[data[i] for i in group])
if p_value > 0.5:
    print("H0 is accepting, There is a relationship between two columns we're comparing")
else:
    print("H0 is rejected , There is no relationship between two columns we're comparing")

H0 is rejected , There is no relationship between two columns we're comparing


In [17]:
group=df.Gender.unique()
data={}
for i in group:
    data[i]=df.AnnualSalary[df.Gender==i]
f_value,p_value=stats.f_oneway(*[data[i] for i in group])
if p_value > 0.5:
    print("H0 is accepting, There is a relationship between two columns we're comparing")
else:
    print("H0 is rejected , There is no relationship between two columns we're comparing")

H0 is rejected , There is no relationship between two columns we're comparing


In [18]:
group=df.Purchased.unique()
data={}
for i in group:
    data[i]=df.Age[df.Gender==i]
f_value,p_value=stats.f_oneway(*[data[i] for i in group])
if p_value > 0.5:
    print("H0 is accepting, There is a relationship between two columns we're comparing")
else:
    print("H0 is rejected , There is no relationship between two columns we're comparing")

H0 is rejected , There is no relationship between two columns we're comparing


c:\Users\Saravanan\OneDrive\Desktop\Dataset\.venv\Lib\site-packages\scipy\stats\_stats_py.py:4102: DegenerateDataWarning: at least one input has length 0
  if _f_oneway_is_too_small(samples):


In [19]:
group=df.Purchased.unique()
data={}
for i in group:
    data[i]=df.AnnualSalary[df.Gender==i]
f_value,p_value=stats.f_oneway(*[data[i] for i in group])
if p_value > 0.5:
    print("H0 is accepting, There is a relationship between two columns we're comparing")
else:
    print("H0 is rejected , There is no relationship between two columns we're comparing")

H0 is rejected , There is no relationship between two columns we're comparing


# Data Visualization

In [21]:
fig = px.imshow(df.corr(numeric_only=True), color_continuous_scale= 'greens', title = "Correlation Matrix")
fig.show()

In [22]:
fig = px.sunburst(df,path=['Purchased', 'Gender'],values='Age')
fig.show()

In [23]:
fig = px.sunburst(df,path=['Gender', 'Purchased'],
                    values='AnnualSalary')
fig.show()

# Preprocessing

In [24]:
df.Gender.unique()

array(['Male', 'Female'], dtype=object)

In [25]:
df["Gender"]=df["Gender"].map({"Male":10,"Female":11})
df

Gender  Age  AnnualSalary  Purchased
User ID                                      
385          10   35         20000          0
681          10   40         43500          0
353          10   49         74000          0
895          10   40        107500          1
661          10   25         79000          0
...         ...  ...           ...        ...
863          10   38         59000          0
800          11   47         23500          0
407          11   28        138500          1
299          11   48        134000          1
687          11   44         73500          0

[1000 rows x 4 columns]

# Machine Learning

## Supervised Regression - Linear regression



x1 = Independent

y = Dependent

x = sm.add_constant(x1)

model = sm.OLS(y,x)

result = model.fit()

result.summary() #

yhat = result.params[0]+result.params[1]* x1

In [33]:
y=df["Purchased"]
x1=df["Gender"]
x=sm.add_constant(x1)
result_log = sm.Logit(y,x).fit()
result_log.pred_table()

Optimization terminated successfully.
         Current function value: 0.672699
         Iterations 4


array([[598.,   0.],
       [402.,   0.]])

In [35]:
cm_df=pd.DataFrame(result_log.pred_table())
cm=np.array(cm_df)
acc=(cm[0,0]+cm[1,1])/cm.sum()
print(acc*100)

59.8


In [37]:
cm_df=pd.DataFrame(result_log.pred_table())
cm_df.columns=["predicted 0","predicted 1"]
cm_df=cm_df.rename(index={0:"actual 0",1:"actual 1"})
cm_df

predicted 0  predicted 1
actual 0        598.0          0.0
actual 1        402.0          0.0

In [38]:
result_log.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:              Purchased   No. Observations:                 1000
Model:                          Logit   Df Residuals:                      998
Method:                           MLE   Df Model:                            1
Date:                Sat, 27 Apr 2024   Pseudo R-squ.:                0.001655
Time:                        14:48:49   Log-Likelihood:                -672.70
converged:                       True   LL-Null:                       -673.81
Covariance Type:            nonrobust   LLR p-value:                    0.1353
==============================================================================
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const         -2.4273      1.363     -1.781      0.075      -5.098       0.243
Gender         0.1930      0.129      1.492      0.136      -0.060       0.446
==============================================================================
"""

In [39]:

y=df["Purchased"]
x1=df["Age"]
x=sm.add_constant(x1)
result_log = sm.Logit(y,x).fit()
cm_df=pd.DataFrame(result_log.pred_table())
cm=np.array(cm_df)
acc=(cm[0,0]+cm[1,1])/cm.sum()
print(acc*100)
result_log.summary()
     

Optimization terminated successfully.
         Current function value: 0.443184
         Iterations 7
81.10000000000001


<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:              Purchased   No. Observations:                 1000
Model:                          Logit   Df Residuals:                      998
Method:                           MLE   Df Model:                            1
Date:                Sat, 27 Apr 2024   Pseudo R-squ.:                  0.3423
Time:                        14:49:50   Log-Likelihood:                -443.18
converged:                       True   LL-Null:                       -673.81
Covariance Type:            nonrobust   LLR p-value:                2.557e-102
==============================================================================
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const         -7.9665      0.500    -15.923      0.000      -8.947      -6.986
Age            0.1827      0.012     15.579      0.000       0.160       0.206
==============================================================================
"""

In [40]:
y=df["Purchased"]
x1=df["AnnualSalary"]
x=sm.add_constant(x1)
result_log = sm.Logit(y,x).fit()
cm_df=pd.DataFrame(result_log.pred_table())
cm=np.array(cm_df)
acc=(cm[0,0]+cm[1,1])/cm.sum()
print(acc*100)
result_log.summary()

Optimization terminated successfully.
         Current function value: 0.604682
         Iterations 5
77.10000000000001


<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:              Purchased   No. Observations:                 1000
Model:                          Logit   Df Residuals:                      998
Method:                           MLE   Df Model:                            1
Date:                Sat, 27 Apr 2024   Pseudo R-squ.:                  0.1026
Time:                        14:50:36   Log-Likelihood:                -604.68
converged:                       True   LL-Null:                       -673.81
Covariance Type:            nonrobust   LLR p-value:                 6.380e-32
================================================================================
                   coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------
const           -2.1563      0.178    -12.090      0.000      -2.506      -1.807
AnnualSalary   2.36e-05   2.18e-06     10.842      0.000    1.93e-05    2.79e-05
================================================================================
"""

In [41]:
y=df["Purchased"]
x1=df[["Gender", "Age"]]
x=sm.add_constant(x1)
result= sm.Logit(y,x).fit()
cm_df=pd.DataFrame(result.pred_table())
cm=np.array(cm_df)
acc=(cm[0,0]+cm[1,1])/cm.sum()
print(acc*100)
result_log.summary()

Optimization terminated successfully.
         Current function value: 0.442974
         Iterations 7
81.10000000000001


<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:              Purchased   No. Observations:                 1000
Model:                          Logit   Df Residuals:                      998
Method:                           MLE   Df Model:                            1
Date:                Sat, 27 Apr 2024   Pseudo R-squ.:                  0.1026
Time:                        14:56:46   Log-Likelihood:                -604.68
converged:                       True   LL-Null:                       -673.81
Covariance Type:            nonrobust   LLR p-value:                 6.380e-32
================================================================================
                   coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------
const           -2.1563      0.178    -12.090      0.000      -2.506      -1.807
AnnualSalary   2.36e-05   2.18e-06     10.842      0.000    1.93e-05    2.79e-05
================================================================================
"""

In [44]:
y=df['Purchased']
x1=df[['AnnualSalary','Age']]
x=sm.add_constant(x1)
result=sm.Logit(y,x).fit()
cm_df=pd.DataFrame(result.pred_table())
cm=np.array(cm_df)
acc=(cm[0,0]+cm[1,1])/cm.sum()
print(acc*100)
result_log.summary()

Optimization terminated successfully.
         Current function value: 0.372966
         Iterations 7
82.8


<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:              Purchased   No. Observations:                 1000
Model:                          Logit   Df Residuals:                      997
Method:                           MLE   Df Model:                            2
Date:                Sat, 27 Apr 2024   Pseudo R-squ.:                  0.1031
Time:                        15:00:43   Log-Likelihood:                -604.38
converged:                       True   LL-Null:                       -673.81
Covariance Type:            nonrobust   LLR p-value:                 6.983e-31
================================================================================
                   coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------
const           -3.2890      1.464     -2.246      0.025      -6.159      -0.419
Gender           0.1084      0.139      0.780      0.435      -0.164       0.381
AnnualSalary  2.351e-05   2.18e-06     10.786      0.000    1.92e-05    2.78e-05
================================================================================
"""

In [45]:
y=df["Purchased"]
x1=df[["Age", "AnnualSalary"]]
x=sm.add_constant(x1)
result_log = sm.Logit(y,x).fit()
cm_df=pd.DataFrame(result_log.pred_table())
cm=np.array(cm_df)
acc=(cm[0,0]+cm[1,1])/cm.sum()
print(acc*100)
result_log.summary()

Optimization terminated successfully.
         Current function value: 0.372966
         Iterations 7
82.8


<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:              Purchased   No. Observations:                 1000
Model:                          Logit   Df Residuals:                      997
Method:                           MLE   Df Model:                            2
Date:                Sat, 27 Apr 2024   Pseudo R-squ.:                  0.4465
Time:                        15:00:54   Log-Likelihood:                -372.97
converged:                       True   LL-Null:                       -673.81
Covariance Type:            nonrobust   LLR p-value:                2.205e-131
================================================================================
                   coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------
const          -11.8442      0.769    -15.399      0.000     -13.352     -10.337
Age              0.2161      0.015     14.520      0.000       0.187       0.245
AnnualSalary  3.325e-05   3.21e-06     10.344      0.000     2.7e-05    3.96e-05
================================================================================
"""

In [46]:

y=df["Purchased"]
x1=df[["Gender", "Age", "AnnualSalary"]]
x=sm.add_constant(x1)
result_log = sm.Logit(y,x).fit()
cm_df=pd.DataFrame(result_log.pred_table())
cm=np.array(cm_df)
acc=(cm[0,0]+cm[1,1])/cm.sum()
print(acc*100)
result_log.summary()
     

Optimization terminated successfully.
         Current function value: 0.371479
         Iterations 7
83.6


<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:              Purchased   No. Observations:                 1000
Model:                          Logit   Df Residuals:                      996
Method:                           MLE   Df Model:                            3
Date:                Sat, 27 Apr 2024   Pseudo R-squ.:                  0.4487
Time:                        15:01:07   Log-Likelihood:                -371.48
converged:                       True   LL-Null:                       -673.81
Covariance Type:            nonrobust   LLR p-value:                9.794e-131
================================================================================
                   coef    std err          z      P>|z|      [0.025      0.975]
--------------------------------------------------------------------------------
const           -8.6760      1.974     -4.395      0.000     -12.545      -4.807
Gender          -0.3184      0.186     -1.716      0.086      -0.682       0.045
Age              0.2195      0.015     14.471      0.000       0.190       0.249
AnnualSalary   3.37e-05   3.23e-06     10.426      0.000    2.74e-05       4e-05
================================================================================
"""

'''
        Dep. Variable:	Purchased

Gender                            59.8
Age                               81.1
AnnualSalary                      77.1

Gender, Age                       81.1
Gender, AnnualSalary              77.2
Age, AnnualSalary                 82.8

Gender, Age, AnnualSalary         83.6
'''

In [47]:
fig=px.scatter(df, x="Age",y="Purchased", color="Gender")
fig.show()
     